File: validate_workflows.py
import yaml
import os
import sys

def validate_yaml_files(directory):
    errors = []
    valid_files = []
    
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith(('.yml', '.yaml')):
                file_path = os.path.join(root, file)
                try:
                    with open(file_path, 'r', encoding='utf-8') as f:
                        yaml.safe_load(f)
                    valid_files.append(file_path)
                    print(f"✅ {file_path}: Valid YAML")
                except yaml.YAMLError as e:
                    errors.append(f"{file_path}: {e}")
                    print(f"❌ {file_path}: YAML Error - {e}")
                except Exception as e:
                    errors.append(f"{file_path}: Unexpected error: {e}")
                    print(f"❌ {file_path}: Error - {e}")
    
    return valid_files, errors

def main():
    workflow_dir = '.github/workflows'
    print(f"Validating YAML files in {workflow_dir}...\n")
    
    valid_files, errors = validate_yaml_files(workflow_dir)
    
    print(f"\n{'='*60}")
    print(f"Validation Summary:")
    print(f"{'='*60}")
    print(f"✅ Valid files: {len(valid_files)}")
    print(f"❌ Errors found: {len(errors)}")
    
    if errors:
        print(f"\n❌ Errors:")
        for error in errors:
            print(f"  - {error}")
        sys.exit(1)
    else:
        print(f"\n✅ All YAML files are valid!")
        sys.exit(0)

if __name__ == "__main__":
    main()

File: validate-workflows.js
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

// Get all workflow files
const workflowDir = path.join(__dirname, '.github', 'workflows');
const workflowFiles = [];

function findFiles(dir, extension) {
  const files = fs.readdirSync(dir);
  files.forEach(file => {
    const filePath = path.join(dir, file);
    if (fs.statSync(filePath).isDirectory()) {
      findFiles(filePath, extension);
    } else if (extension.includes(path.extname(file))) {
      workflowFiles.push(filePath);
    }
  });
}

findFiles(workflowDir, ['.yml', '.yaml']);

console.log(`Found ${workflowFiles.length} workflow files to validate\n`);

let validFiles = 0;
let invalidFiles = 0;

workflowFiles.forEach(filePath => {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    yaml.load(content);
    console.log(`✅ ${path.relative(__dirname, filePath)} - Valid YAML`);
    validFiles++;
  } catch (error) {
    console.log(`❌ ${path.relative(__dirname, filePath)} - Invalid YAML: ${error.message}`);
    invalidFiles++;
  }
});

console.log(`\nValidation Summary:`);
console.log(`Valid files: ${validFiles}`);
console.log(`Invalid files: ${invalidFiles}`);
console.log(`Total files: ${workflowFiles.length}`);
